src/thunderbots/software/util/visualizer_messenger/visualizer_messenger.cpp

```diff
@@ -0,0 +1,171 @@
+                                      double radius, double theta_start, double theta_end,
+                                      DrawStyle draw_style, DrawTransform draw_transform)
+    {
+        ShapeMsg new_shape;
+        new_shape.type = "arc";
+        new_shape.data.clear();
+        new_shape.data.push_back(cx);
+        new_shape.data.push_back(cy);
+        new_shape.data.push_back(radius);
+        new_shape.data.push_back(theta_start);
+        new_shape.data.push_back(theta_end);
+
+        applyDrawStyleToMsg(new_shape, draw_style);
+        applyDrawTransformToMsg(new_shape, draw_transform);
+        addShapeToLayer(layer, new_shape);
+    }
+
+    void VisualizerMessenger::drawLine(const std::string& layer, double x1, double y1,
+                                       double x2, double y2, DrawStyle draw_style,
+                                       DrawTransform draw_transform)
+    {
+        ShapeMsg new_shape;
+        new_shape.type = "line";
+        new_shape.data.clear();
+        new_shape.data.push_back(x1);
+        new_shape.data.push_back(y1);
+        new_shape.data.push_back(x2);
+        new_shape.data.push_back(y2);
+
+        applyDrawStyleToMsg(new_shape, draw_style);
+        applyDrawTransformToMsg(new_shape, draw_transform);
+        addShapeToLayer(layer, new_shape);
+    }
+
+    void VisualizerMessenger::buildLayers()
+    {
+        // TODO: #268 list these existing layer names elsewhere where it's more obvious
+        std::vector<std::string> layer_names = std::vector<std::string>(
+            {"field", "ball", "ball_vel", "fiendly_robot", "fiendly_robot_vel",
+             "enemy_robot", "enemy_robot_vel", "nav", "rule", "passing", "test", "misc"});
+
+        for (std::string layer_name : layer_names)
+        {
+            LayerMsg new_layer_msg;
+            new_layer_msg.layer_name = layer_name;
+            m_layers_name_to_msg_map.insert(
+                std::pair<std::string, LayerMsg>(layer_name, new_layer_msg));
+        }
+    }
+
+    void VisualizerMessenger::applyDrawStyleToMsg(ShapeMsg& shape_msg, DrawStyle& style)
+    {
+        shape_msg.fill          = style.fill;
+        shape_msg.stroke        = style.stroke;
+        shape_msg.stroke_weight = style.stroke_weight;
+    }
+
+    void VisualizerMessenger::applyDrawTransformToMsg(ShapeMsg& shape_msg,
+                                                      DrawTransform& transform)
+    {
+        shape_msg.transform_rotation = transform.rotation;
+        shape_msg.transform_scale    = transform.scale;
+    }
+
+    void VisualizerMessenger::addShapeToLayer(const std::string& layer, ShapeMsg& shape)
+    {
+        if (m_layers_name_to_msg_map.find(layer) != m_layers_name_to_msg_map.end())
+        {
+            m_layers_name_to_msg_map[layer].shapes.emplace_back(shape);
+        }
+        else
+        {
+            LOG(WARNING) << "Referenced layer (" << layer << ") is undefined\n";
```

Comment:
nit (so feel free to leave this as-is); but I don't think we need `\n` at the end of the log message when using our logger.

Reply:
I think you do need it in this case since you don't have `<< std::endl;` at the end instead

Reply:
ah, fair enough